### PR TITLE
Add TSX Switch for Haswell/Broadwell in Debug settings

### PR DESCRIPTION
**Pull request by NeoRetro**
New description: This PR adds a switch in the debug tab of the settings to enable TSX functions on Broadwell and unpatched Haswell CPUs.
Huge Credits @Megamouse for the setting UI and config implementation - I would have taken ages to implement it all by myself.

[Old description: The current implementation to activate TSX checks for MPX extensions, which were introduced with Skylake. Therefore Broadwell is currently locked out, even though having TSX implementations.

Haswell is still locked out, because TSX instruction set is broken there.]

### CODE REVIEW

**Comment by Nekotekina** [comment]
What CPU models you want to whitelist?
Many Broadwell CPUs have similarly broken TSX support.

**Comment by NeoRetro** [comment]
With this change Broadwell CPUs like the i7-5775C and also Broadwell-E like the i7-6800K are able to use TSX now.
Broadwell and Skylake indeed had TSX bugs at launch, which were solved by Intel Microcode revisions.
So to Blacklist all potentially TSX-corrupted we would also need to blacklist Skylake CPUs - which is not done at the moment.
This PR just treats Broadwell just the same as Skylake.

**Comment by jaysharp29** [comment]
I can also confirm a huge speed up using tsx on a 6850k.  Demon souls went from 20-30 fps to faster than an actual ps3.  Played for 2 hours and not a single crash.  Shader compilation was faster as well.

A manual check seems ideal.

**Comment by danilaml** [comment]
Aren't broadwells (desktop ones) simply Haswell on 14nm? I.e. same "broken" TSX-set (which could lead to some locked state in rare circumstances but is probably fine in general).

**Comment by NeoRetro** [comment]
Actually these have hardware fixes for the deep hardware bugs in Haswell that pushed Intel to disable TSX with newer micocode revisions.
A good example: Because of the Spectre/Meltdown vulnerabilities Intel was forced to release microcode updates for almost every CPU they sold starting something around 2009... these new microcode updates disable TSX on Haswell, while still leaving them enabled for Broadwell.

To the topic of having a switch to enable TSX on Haswell: As unfortunate it is for people owning Haswell (I also have one PC running it...) my opinion is that the chances of breaking RPCS and also your operating system (Windows/Linux) is too much of a risk to put it into the options menu. Maybe some additions in some config.xml but i would advise against some switch in the options menu...

**Comment by Cereal-Killa** [comment]
I'm willing to do some work on that if people agree that the commit will be accepted later.

pseudo code (just a supposition, don't take it seriously...):

`
bool utils::has_rtm()
{
	// Check RTM and filter out TSX on Haswell CPUs
	static const bool g_value = get_cpuid(0, 0)[0] >= 0x7 && (get_cpuid(7, 0)[1] & 0x800) == 0x800 && rtm_enabled());
	return g_value;
}

bool utils:rtm_enabled()
{
        return jit_compiler::cpu("") != "haswell" || g_cfg.core.broken_tsx_enabled;
}
`

**Comment by Nekotekina** [comment]
Better implementation would include checking particular CPU models, by the name retrieved from CPUID. It's not that hard although it may be tricky. Proposed implementation uses LLVM tools which is unsafe and undesirable.

**Comment by danilaml** [comment]
Do CPUs report TSX as supported even after microcode update that disables it?

**Comment by Nekotekina** [comment]
Doesn't mean people use the latest microcode. They run RPCS3 and may have various fun, like ruined performance.

File: Utilities/sysinfo.cpp [hunk 1]
```diff
@@ -33,8 +33,16 @@ bool utils::has_avx2()
 
 bool utils::has_rtm()
 {
-	// Check RTM and MPX extensions in order to filter out TSX on Haswell CPUs
-	static const bool g_value = get_cpuid(0, 0)[0] >= 0x7 && (get_cpuid(7, 0)[1] & 0x4800) == 0x4800;
+	// Check RTM
+	static bool g_value = get_cpuid(0, 0)[0] >= 0x7 && (get_cpuid(7, 0)[1] & 0x800) == 0x800;
+	if (g_value) {
```
Comment (Megamouse) [inline Utilities/sysinfo.cpp]: we put the `{` into a new line

Comment (NeoRetro) [reply Utilities/sysinfo.cpp]: Thank you! Fixed it.

**Comment by tabnk** [comment]
Intel 4590 Processor does not support TSX, otherwise it should help some performance gain .

**Comment by Shoegzer** [comment]
Interestingly, Kaby Lake CPUs such as the i7-7820HK [support TSX-NI](https://ark.intel.com/products/97464/Intel-Core-i7-7820HK-Processor-8M-Cache-up-to-3_90-GHz) but RPCS3's UI does not report TSX.  Is there a reason?  I thought anything past Broadwell would not be a problem.

**Comment by Gamedev1909** [comment]
Its been 3 days now testing tsx on 4790k and ive seen 4690k as well all have had no issues and major speed boosts up to 40%

**Comment by Megamouse** [comment]
You should not use Qt stuff in that file.
Also I would strongly advice to use a debug config setting instead of this approach.
the QMessageBox could be displayed whenever enabling the checkbox.
Also the checkbox can be hidden on unaffected CPUs.

**Comment by Syrusse** [comment]
So I'm the crazy guy who want TSX but have no knowledge in programming, I'm not understanding very well what's going so, I just want to understand if TSX is going to be implemented for us the pleb who can't afford a new CPU or not ? I understand the risk and everything but from what I know, everyone using Windows 10 (even with no updates at all and with the very first build) is running with the microcode revision 1E (so after the disabling of TSX) I also understand that devs doesn't want to add this feature as it could lead to serious problems ? But as far as I know no one was concerned and furthermore we could simply add a huge warning message for the persons who are using this feature and where we could tell them that it could cause serious problem and that you should not enable that feature if you don't know what you're doing...
In fact, the odds that someone is using TSX with an haswell CPUs with an old microcode (that would mean he either deleted the two mcupdate files or he would NEVER download any updates in W7/8.1) are near 0.
So please devs, I beg you, for the sake of peoples who can't build RPCS3 (I couldn't build the latest one and Appveyor returned me a error that I couldn't fix of course) please, add this option for us, I can't afford to pay anything else for my PC, if I could afford my I7 4770, it was only because I got my mobo (bent pins), my RAM, my 2 HDD for free.

**Comment by NeoRetro** [comment]
As I'm currently on a 1 week holiday, an implementation of a debug setting on my little notebook is quite tricky... I can open a new PR for this debug setting for Haswell when I'm back.
@Nekotekina Do you think this PR can be merged like this?

**Comment by Megamouse** [comment]
Just keep the PR open and push a new commit when you come back.
Users who can't wait will still have this appveyor build to play with.
😄

**Comment by Cereal-Killa** [comment]
@NeoRetro We'll be waiting, thank you.

**Comment by Megamouse** [comment]
I was bored so I figured I could implement it, too. In case you want to cherrypick it, just say so.

**Comment by NeoRetro** [comment]
@Megamouse If you have time for it feel free to do it 😄

**Comment by Megamouse** [comment]
I just need to fix a static check, then I can push it for you.

File: Utilities/sysinfo.cpp [hunk 2]
```diff
@@ -33,8 +33,25 @@ bool utils::has_avx2()
...
+	}
+	else
+	{
+		// Get CPU Model
+		int Model = (get_cpuid(1, 0)[0] >> 4) & 0xf;
+		Model += ((get_cpuid(1, 0)[0] >> 16) & 0xf) << 4;
+		
+		// If CPU is Haswell, disable TSX (for now)
+		g_value = Model != 0x3c && Model != 0x3f && Model != 0x45 && Model != 0x46;
+		RTMChecked = !RTMChecked;
```
Comment (kd-11) [inline Utilities/sysinfo.cpp]: Use snake_case for consistency with the rest of the file and compliance with the emulator coding style guidelines. Minor nit really, only `Model` and `RTMChecked` need to be changed

**Comment by Megamouse** [comment]
Feel free to cherry-pick this commit:
https://github.com/Megamouse/rpcs3/commit/7504e07cfb96d476e4194239650d55a8fdc2eea6

It adds a checkbox to the debug tab in the gui settings that allows you to enable TSX on Haswell and Broadwell CPUs. The checkbox is hidden if you have any other CPU. It is disabled by default.

**Comment by Zangetsu38** [comment]
Stop lol, rebase and squash all, here is dirty, and no using website for edit code

**Comment by Gamedev1909** [comment]
Dont ruin things on this you gotta calm down

File: rpcs3/rpcs3qt/emu_settings.h [hunk 3]
```diff
@@ -197,6 +198,7 @@ public Q_SLOTS:
 		{ PPUDebug,                 { "Core", "PPU Debug"}},
 		{ SPUDebug,                 { "Core", "SPU Debug"}},
 		{ MaxLLVMThreads,           { "Core", "Max LLVM Compile Threads"}},
+		{ EnableTSX,				{ "Core", "Enable TSX on Haswell/Broadwell"}},
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/emu_settings.h]: use spaces instead of tabs here. the indentation is broken.

File: rpcs3/Emu/System.cpp [hunk 4]
```diff
@@ -439,6 +443,20 @@ void Emulator::Load(bool add_only)
...
 
 		LOG_NOTICE(LOADER, "Used configuration:\n%s\n", g_cfg.to_string());
+		
+		// Set RTM usage
+		g_use_rtm = utils::has_rtm() && (utils::is_haswell_or_broadwell() == utils::ModelNotHaswellBroadwell || g_cfg.core.enable_TSX);
+		if (utils::has_rtm())
+		{
+			if (g_use_rtm)
+			{
+				LOG_NOTICE(GENERAL, "TSX enabled");
```
Comment (Megamouse) [inline rpcs3/Emu/System.cpp]: these are more or less useless, since you can see this on the log anyway

File: Utilities/sysinfo.cpp [hunk 5]
```diff
@@ -51,6 +50,22 @@ bool utils::has_xop()
 	return g_value;
 }
 
+int utils::is_haswell_or_broadwell()
+{
+	static const int g_model = ((get_cpuid(1, 0)[0] >> 4) & 0xf) + (((get_cpuid(1, 0)[0] >> 16) & 0xf) << 4);
+	//static const bool g_value = g_model == 0x3c || g_model == 0x3f || g_model == 0x45 || g_model == 0x46 || // Haswell
```
Comment (Megamouse) [inline Utilities/sysinfo.cpp]: you could modify this to use an int with a ternary expression

File: Utilities/sysinfo.h [hunk 6]
```diff
@@ -40,6 +40,15 @@ namespace utils
 	bool has_512();
 
 	bool has_xop();
+	
+	int is_haswell_or_broadwell();
+	
+	enum haswell_broadwell
```
Comment (TGEnigma) [inline Utilities/sysinfo.h]: There is really no need to have an enum for this. Just make seperate check methods for haswell and broadwell

File: rpcs3/Emu/System.cpp [hunk 7]
```diff
@@ -439,6 +443,20 @@ void Emulator::Load(bool add_only)
...
+			}
+			else
+			{
+				LOG_NOTICE(GENERAL, "TSX disabled");
```
Comment (TGEnigma) [inline rpcs3/Emu/System.cpp]: Same goes for this

File: rpcs3/rpcs3qt/settings_dialog.cpp [hunk 8]
```diff
@@ -960,6 +961,58 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> guiSettings, std:
...
+			if (utils::is_haswell_or_broadwell() == utils::ModelHaswell)
+			{
+				if (checked && QMessageBox::No == QMessageBox::critical(this, tr("Haswell TSX Warning"), tr(
+					R"(
+					<p style="white-space: nowrap;">
+						RPCS3 has detected you are using TSX functions on a Haswell CPU.<br>
+						Intel has deactivated these functions in newer Microcode revisions, since they can lead to unpredicted behaviour.<br>
+						That means using TSX may break games or even <font color="red"><b>damage</b></font> your data.<br>
+						We recommend to disable this feature and update your computer BIOS.<br><br>
+						Do you wish to use TSX anyway?
```
Comment (TGEnigma) [inline rpcs3/rpcs3qt/settings_dialog.cpp]: This line is different from the Broadwell version for no good reason

File: Utilities/sysinfo.cpp (hunk 5 above)
Comment (TGEnigma) [reply Utilities/sysinfo.cpp]: that would take up an unnecessary amount of space and hurt readability considering how many conditions there are being checked

Comment (TGEnigma) [reply Utilities/sysinfo.cpp]: that would take up an unnecessary amount of space and hurt readability considering how many conditions there are being checked

File: rpcs3/rpcs3qt/emu_settings.h (hunk 3 above)
Comment (NeoRetro) [reply rpcs3/rpcs3qt/emu_settings.h]: In Visual Studio it looked good - I'll fix that.

File: rpcs3/rpcs3qt/settings_dialog.cpp (hunk 8 above)
Comment (NeoRetro) [reply rpcs3/rpcs3qt/settings_dialog.cpp]: Difference is: Haswell TSX is removed by Intel in new Microcode/BIOS versions. Enable TSX on Haswell is unsafe.
Broadwell TSX was fixed by Intel in new Microcode/BIOS versions. Enable TSX on Broadwell is safe.

File: rpcs3/Emu/System.cpp (hunk 4 above)
Comment (NeoRetro) [reply rpcs3/Emu/System.cpp]: I'll change it so the log is only displayed on Haswell and Broadwell CPUs.

File: Utilities/sysinfo.cpp [hunk 9]
```diff
@@ -33,8 +33,7 @@ bool utils::has_avx2()
 
 bool utils::has_rtm()
 {
-	// Check RTM and MPX extensions in order to filter out TSX on Haswell CPUs
-	static const bool g_value = get_cpuid(0, 0)[0] >= 0x7 && (get_cpuid(7, 0)[1] & 0x4800) == 0x4800;
+	static bool g_value = get_cpuid(0, 0)[0] >= 0x7 && (get_cpuid(7, 0)[1] & 0x800) == 0x800;
```
Comment (Megamouse) [inline Utilities/sysinfo.cpp]: you can use static const bool here again

File: rpcs3/rpcs3qt/settings_dialog.ui [hunk 10]
```diff
@@ -1725,6 +1725,13 @@
               </property>
              </widget>
             </item>
+            <item>
+			          <widget class="QCheckBox" name="enableTSX">
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/settings_dialog.ui]: why did you add three tabs to all of these already (with spaces) indented lines

File: Utilities/sysinfo.cpp (hunk 9 above)
Comment (NeoRetro) [reply Utilities/sysinfo.cpp]: Yes, you're right - changed it

File: rpcs3/rpcs3qt/settings_dialog.ui (hunk 10 above)
Comment (NeoRetro) [reply rpcs3/rpcs3qt/settings_dialog.ui]: Don't know why this happened lol

File: rpcs3/rpcs3qt/settings_dialog.cpp (hunk 8 above)
Comment (TGEnigma) [reply rpcs3/rpcs3qt/settings_dialog.cpp]: Referring to the very last line

File: Utilities/sysinfo.cpp [hunk 11]
```diff
@@ -51,6 +50,20 @@ bool utils::has_xop()
 	return g_value;
 }
 
+bool utils::is_haswell()
+{
+	static const int g_model = ((get_cpuid(1, 0)[0] >> 4) & 0xf) + (((get_cpuid(1, 0)[0] >> 16) & 0xf) << 4);
```
Comment (TGEnigma) [inline Utilities/sysinfo.cpp]: the assignment for g_model is the same for both haswell & broadwell cases. Refactor into seperate method, eg utils::get_cpu_model()

Comment (NeoRetro) [reply Utilities/sysinfo.cpp]: Restructured it.

File: rpcs3/Emu/System.cpp (hunk 7 above)
Comment (NeoRetro) [reply rpcs3/Emu/System.cpp]: screen tsx

Comment (NeoRetro) [reply rpcs3/Emu/System.cpp]: Was writing the same above: For Haswell/Broadwell Users the "| TSX" will appear if the CPU has the capabilities, regardless if it's enabled or not. Therefore this Info log will only show for them.

File: rpcs3/rpcs3qt/settings_dialog.cpp (hunk 8 above)
Comment (NeoRetro) [reply rpcs3/rpcs3qt/settings_dialog.cpp]: fixed it.

File: rpcs3/Emu/System.h [hunk 12]
```diff
@@ -311,7 +311,8 @@ struct cfg_root : cfg::node
 		cfg::_int<0, 16> spu_delay_penalty{this, "SPU delay penalty", 3}; //Number of milliseconds to block a thread if a virtual 'core' isn't free
 		cfg::_bool spu_loop_detection{this, "SPU loop detection", true}; //Try to detect wait loops and trigger thread yield
 		cfg::_bool spu_shared_runtime{this, "SPU Shared Runtime", true}; // Share compiled SPU functions between all threads
-		cfg::_bool enable_TSX{ this, "Enable TSX on Haswell/Broadwell", false }; // Enable TSX on Haswell CPUs. This should be used carefully
+    cfg::_enum<spu_block_size_type> spu_block_size{this, "SPU Block Size"};
```
Comment (Megamouse) [inline rpcs3/Emu/System.h]: you messed with up the conflict

File: rpcs3/Emu/System.cpp (hunk 7 above)
Comment (Megamouse) [reply rpcs3/Emu/System.cpp]: If you insist on adding this, you should at least use a LOG_WARNING instead of a LOG_NOTICE
:)

**Comment by Syrusse** [comment]
Finally ! Thank you very much ! :3

**Comment by NeoRetro** [comment]
Greatly simplified the PR now.

**Comment by incognitoh** [comment]
It seems Haswell TSX is broken with this latest merge. In fact, Haswell TSX has not been working correctly on any build for the last 3 days or so. Before that, I would manually hx edit to enable TSX and it all worked fine with last week's builds but something changed a few days ago. But now, I tried 10 games and 8 of them lock up right away with TSX enabled. Without TSX enabled, they all work fine.

**Comment by Syrusse** [comment]
Man, I exactly have the same issue, I was previously hex editing and everything was fine, until I tried the super big update, and, nothing ! It wont shows up :o 
So I'll try this build and give me results if I can figure out how I can download it

**Comment by Fiodorwellfme** [comment]
Ooooooyea, I've joined the github master race!
Guess we're all in the same boat :(

**Comment by NeoRetro** [comment]
@incognitoh can you find out the exact PR that broke TSX for your CPU?

**Comment by Syrusse** [comment]
@Asinin3 You're talking about the method when we replace have_rts by always true ?
If it's this, it doesn't work anymore as compiling the emulator result to an error :c

**Comment by Fiodorwellfme** [comment]
I'm building it through appveyor to see if it is related to using the hex editor.

**Comment by Asinin3** [comment]
No, I'm talking about this specifically https://www.reddit.com/r/rpcs3/comments/8emx9u/up_to_40_fps_boost_on_some_haswell_cpus/dyd9uur/ which you replied to @Syrusse

**Comment by incognitoh** [comment]
Asinin3, No using both hexedit method and the latest merge here (without any hexedit) https://ci.appveyor.com/project/rpcs3/rpcs3/build/0.0.5-92865b32/artifacts both result in quick lockup. But with TSX disabled in the debug menu, it works fine.

NeoRetro, I will try to narrow down when hexedits stopped working correctly. Thanks

**Comment by Syrusse** [comment]
@Asinin3 Lol, I'm the same guy who created that post ^^
Yeah, that method definitely doesn't work anymore, it simply consist to replace a value by always true, unfortunately, when I'm repeating the same steps than before, appveyor can't compile the emulator for some reasons, this is why I switched to "hex editing" and now, with one of the latest build, h"hex editing" doesn't also work, so NeoRetro is our last hope :p

**Comment by TGEnigma** [comment]
Keep in mind that if you recently updated Windows 10, you will have to re-patch your cpu microcode & remove the microcode updater dlls again.

**Comment by Cereal-Killa** [comment]
What a coincidence, I was reading about TSX on the Intel website in the beginning of this week and that's when I decided to look for some of the intrinsics on the repository and found that pull request which removed the "_xbegin" from the code...

https://software.intel.com/en-us/node/695154

**Comment by blottto** [comment]
@incognitoh  I have the same problem is broken TSX now in Haswell I have an I7-4790@4.35

@NeoRetro This is when master builds stopped working with hexedits: https://ci.appveyor.com/project/rpcs3/rpcs3/build/0.0.5-6750/artifacts <-- this build does not work with Haswell TSX

This build is the last master that does work with haswell TSX edit: https://ci.appveyor.com/project/rpcs3/rpcs3/build/0.0.5-6748/artifacts <-- this build works

Hopefully that will help to figure this out.

(btw I am on Windows 7 and have not updated anything)

**Comment by Cereal-Killa** [comment]
Same here, broken now, hopefully someone can find out what the issue is and restore Haswell compatibility.

Made a short video today on my PC of how bennefitial TSX can be on Haswell: https://www.youtube.com/watch?v=jE_OWKBimLQ&feature=youtu.be

**Comment by Nekotekina** [comment]
First TSX support implementation was incompatible with Haswell as well. Only "second" one had been working for some reason.
I suggest trying my PR #4705 because it reduces the use of transactions, but it's nothing but trial and error.

**Comment by Cereal-Killa** [comment]
Just tried #4705, unfortunately the experimental TSX check box from this pull request isn't there, the intro video still stutters a lot without TSX.

**Comment by Nekotekina** [comment]
Someone needs to force enable it in order to test. I can't make a build currently.

**Comment by blottto** [comment]
I activated support for TSX using HxD and it worked again yeaa # 4705
Thanks @Nekotekina

came back to work TSX in haswell

**Comment by Cereal-Killa** [comment]
@blottto can you share the bytes you're patching so I can test here.

**Comment by blottto** [comment]
This is momentary while in the future you can add the option activate tsx tab debug in cpu 4 and 5 generation

using hex editor and replace 25 00 48 00 00 3D 00 48 00 00 with 25 00 08 00 00 3D 00 08 00 00

**Comment by Cereal-Killa** [comment]
Tks @blottto, I patched successfully however both games I had froze the entire emulator at a certain point, had to end the task.

**Comment by Megamouse** [comment]
chuck-norris-pull-request

**Comment by Nekotekina** [comment]
Would still be better to be able to disable it completely

**Comment by Megamouse** [comment]
I think we can add this ourselves in a follow up

File: rpcs3/Json/tooltips.json [hunk 13]
```diff
@@ -37,6 +37,7 @@
 	"debug": {
 		"ppuDebug": "Never use this.",
 		"spuDebug": "Never use this.",
+		"enableTSX": "Force usage of TSX instructions on Haswell/Broadwell CPUs.\nOn older hardware this can lead to system instability, use it with caution.",
```
Comment (Megamouse) [inline rpcs3/Json/tooltips.json]: In order to make this universal change the tooltip to something like:
"Force usage of TSX instructions.\nOn older Haswell or Broadwell CPUs this can lead to system instability, use it with caution."

File: rpcs3/rpcs3qt/emu_settings.h [hunk 14]
```diff
@@ -198,6 +199,7 @@ public Q_SLOTS:
 		{ PPUDebug,                 { "Core", "PPU Debug"}},
 		{ SPUDebug,                 { "Core", "SPU Debug"}},
 		{ MaxLLVMThreads,           { "Core", "Max LLVM Compile Threads"}},
+		{ EnableTSX,                { "Core", "Enable TSX on Haswell/Broadwell"}},
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/emu_settings.h]: In order to make this universal change the name to simply: "Enable TSX"

File: rpcs3/Emu/System.h [hunk 15]
```diff
@@ -314,6 +314,7 @@ struct cfg_root : cfg::node
 		cfg::_bool spu_loop_detection{this, "SPU loop detection", true}; //Try to detect wait loops and trigger thread yield
 		cfg::_bool spu_shared_runtime{this, "SPU Shared Runtime", true}; // Share compiled SPU functions between all threads
 		cfg::_enum<spu_block_size_type> spu_block_size{this, "SPU Block Size"};
+		cfg::_bool enable_TSX{this, "Enable TSX on Haswell/Broadwell", false}; // Enable TSX on Haswell/Broadwell CPUs. This should be used carefully
```
Comment (Megamouse) [inline rpcs3/Emu/System.h]: In order to make this universal change the string to simply: "Enable TSX"

File: rpcs3/Emu/System.cpp [hunk 16]
```diff
@@ -509,6 +511,13 @@ void Emulator::Load(bool add_only)
 #endif
 
 		LOG_NOTICE(LOADER, "Used configuration:\n%s\n", g_cfg.to_string());
+		
+		// Set RTM usage
+		g_use_rtm = utils::has_rtm() && (utils::has_mpx() || g_cfg.core.enable_TSX);
+		if (g_cfg.core.enable_TSX)
```
Comment (Megamouse) [inline rpcs3/Emu/System.cpp]: In order to make this universal you should use:
`if (g_cfg.core.enable_TSX && utils::has_mpx())`
in order to remove the message on supported hardware

File: rpcs3/rpcs3qt/settings_dialog.ui [hunk 17]
```diff
@@ -1729,6 +1729,13 @@
               </property>
              </widget>
             </item>
+            <item>
+             <widget class="QCheckBox" name="enableTSX">
+              <property name="text">
+               <string>Enable TSX on Haswell/Broadwell</string>
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/settings_dialog.ui]: In order to make this universal change the name to simply: "Enable TSX"

File: rpcs3/rpcs3qt/settings_dialog.cpp [hunk 18]
```diff
@@ -973,6 +974,36 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> guiSettings, std:
 
 	xemu_settings->EnhanceCheckBox(ui->spuDebug, emu_settings::SPUDebug);
 	SubscribeTooltip(ui->spuDebug, json_debug["spuDebug"].toString());
+		
+	if (utils::has_rtm() && !utils::has_mpx())
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/settings_dialog.cpp]: In order to make this universal change this to:
`if (utils::has_rtm())`

File: rpcs3/rpcs3qt/settings_dialog.cpp [hunk 19]
```diff
@@ -973,6 +974,36 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> guiSettings, std:
...
+	{
+		xemu_settings->EnhanceCheckBox(ui->enableTSX, emu_settings::EnableTSX);
+		SubscribeTooltip(ui->enableTSX, json_debug["enableTSX"].toString());
+
+		// connect the toogled signal so that the stateChanged signal in EnhanceCheckBox can be prevented
+		connect(ui->enableTSX, &QCheckBox::toggled, [this](bool checked)
+		{
+			if (checked && QMessageBox::No == QMessageBox::critical(this, tr("Haswell/Broadwell TSX Warning"), tr(
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/settings_dialog.cpp]: In order to make this universal change this to:
`if (checked && !utils::has_mpx() && QMessageBox::No == ....`

**Comment by Megamouse** [comment]
also you need to solve the merge conflict

File: rpcs3/Emu/System.cpp [hunk 20]
```diff
@@ -509,6 +511,13 @@ void Emulator::Load(bool add_only)
 #endif
 
 		LOG_NOTICE(LOADER, "Used configuration:\n%s\n", g_cfg.to_string());
+		
+		// Set RTM usage
+		g_use_rtm = utils::has_rtm() && (utils::has_mpx() || g_cfg.core.enable_TSX);
```
Comment (Megamouse) [inline rpcs3/Emu/System.cpp]: In order to make this universal you should use:
`g_use_rtm = utils::has_rtm() && g_cfg.core.enable_TSX`

**Comment by Megamouse** [comment]
My review is not final. It may be wrong. So better not touch it yet
But you need to solve the merge conflict

File: rpcs3/Emu/System.h (hunk 15 above)
Comment (Megamouse) [reply rpcs3/Emu/System.h]: add an enum at the top with `{disabled, enabled, forced}` for this

File: rpcs3/rpcs3qt/settings_dialog.cpp [hunk 21]
```diff
@@ -973,6 +974,36 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> guiSettings, std:
 
 	xemu_settings->EnhanceCheckBox(ui->spuDebug, emu_settings::SPUDebug);
 	SubscribeTooltip(ui->spuDebug, json_debug["spuDebug"].toString());
+		
+	if (utils::has_rtm() && !utils::has_mpx())
+	{
+		xemu_settings->EnhanceCheckBox(ui->enableTSX, emu_settings::EnableTSX);
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/settings_dialog.cpp]: In order to make this universal change this to:
xemu_settings->EnhanceComboBox(ui->enableTSX, emu_settings::EnableTSX);

**Comment by Megamouse** [comment]
due to this PR being kinda lost and forgotten + merge conflict 
I decided to open a proper new PR.
I hope you are fine with it.

https://github.com/RPCS3/rpcs3/pull/4766

**Comment by Megamouse** [comment]
this can be closed now 🥇

**Comment by Nekotekina** [comment]
@NeoRetro Thank you for the contribution. The feature has been merged.
